# Домашнее задание "Теория оптимизации".

## Задание.

При фиксированном seed=42 поразбирайтесь и поэкспериментируйте с параметрами алгоритма дифференциальной эволюции: strategy, popsize, tol, mutation, recombination и updating.

Постройте графики количества итераций оптимизации функции ackley от значения параметра.

In [2]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import dual_annealing
from scipy.optimize import differential_evolution
from scipy.optimize import basinhopping

In [4]:
# Опять мне лично мало входных данных. Вообще мне кажется, что домашние задания были изначально подготовлены для другого
# курса, с другими лекциями, с другим информационным наполнением. Потом мнформационное наполнение лекций сократили, а ДЗ
# не поменяли.
# К какой функции применять алгоритм дифференциальной эволюции? Предположу, что надо взять какую-то (на мое усмотрение) со
# со страницы Википедии (https://en.wikipedia.org/wiki/Test_functions_for_optimization). Мне понравилась из-за ее названия
# функция Three-hump camel function

# Зададим эту функцию:
def Three_hump_camel_fun(x):
    return 2*x[0]**2 - 1.05*x[0]**4 + (x[0]**6)/6 + x[0]*x[1] + x[1]**2

# задаем пределы изменения переменных
bounds = [(-5, 5), (-5, 5)]

# Для нашей верблюжей функции и заданных границ переменных применим функцию differential_evolution с параматерами по 
# умолчанию и seed=42
result0 = differential_evolution(Three_hump_camel_fun, bounds, seed=42)
result0

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 3033
     nit: 100
 success: True
       x: array([0., 0.])

Итак, функция вернула результат:<br>
Оптимум функции равен **0** <br>
Оптимизация завершена успешно<br>
Количество оценок целевых функций **3033**<br>
Количество выполненных итераций **100** <br>
Успех и правда успех. И вообще скоро Новый год! <br>
Оптимум функции достигается в точке **[0, 0]**<br>

In [5]:
# Теперь изменим параметр strategy. По умолчанию он strategy='best1bin'. Укажем, например, strategy='rand2exp'
result1 = differential_evolution(Three_hump_camel_fun, bounds, strategy='rand2exp', seed=42)
result1

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 10563
     nit: 351
 success: True
       x: array([0., 0.])

Здесь сразу оговорюсь: все результаты экспериментов буду сравнивать с result0 - резултатом для параметров по умолчанию<br>
Оптимум функции так же равен **0** <br>
Количество оценок целевых функций увеличелось до **10563**<br>
Количество выполненных итераций увеличелось до **351** <br>
Оптимум функции так же достигается в точке **[0, 0]**<br>

In [17]:
# Теперь изменим параметр popsize - множитель для определения общей численности популяции. 
# По умолчанию он popsize=15. Укажем, например, popsize=75
result2 = differential_evolution(Three_hump_camel_fun, bounds, popsize=39, seed=42)
result2

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 8271
     nit: 105
 success: True
       x: array([0., 0.])

Оптимум функции так же равен **0** <br>
Количество оценок целевых функций увеличелось до **8271**<br>
Количество выполненных итераций увеличелось до **105** <br>
Оптимум функции так же достигается в точке **[0, 0]**<br>

In [16]:
# Укажем, например, popsize=3
result2 = differential_evolution(Three_hump_camel_fun, bounds, popsize=4, seed=42)
result2

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 883
     nit: 109
 success: True
       x: array([0., 0.])

Оптимум функции так же равен **0** <br>
Количество оценок целевых функций сократилось до **883**<br>
Количество выполненных итераций увеличелось до **109** <br>
Оптимум функции так же достигается в точке **[0, 0]**<br>

In [23]:
# Теперь изменим параметр tol - относительный допуск для сходимости. По умолчанию он tol=0.01. 
# Укажем, например, tol=0.0005
result3 = differential_evolution(Three_hump_camel_fun, bounds, tol=0.0005, seed=42)
result3

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 3033
     nit: 100
 success: True
       x: array([0., 0.])

Уменьшение параметра tol в 20 раз не повлияло на результат - такой же как у result0

In [22]:
# Укажем, например, tol=0.8
result3 = differential_evolution(Three_hump_camel_fun, bounds, tol=0.8, seed=42)
result3

     fun: 1.8143565380354957e-13
     jac: array([-8.02564897e-07,  3.86601506e-07])
 message: 'Optimization terminated successfully.'
    nfev: 105
     nit: 2
 success: True
       x: array([-2.88818757e-07,  3.32710131e-07])

Если постепенно увеличивать параметр tol, то по началу так же результат не меняется, но при каком-то достаточно большом значении начинает страдать точность получаемых результатов. Например, при tol=0.8 было выполнено всего 2 итерации и в качестве решений получены значения близкие к нулю, но не равные нулю. 

In [32]:
# Теперь изменим параметр mutation - постоянную мутации. По умолчанию mutation=(0.5, 1). 
# Укажем, например, mutation=(0.1, 0.2)
result4 = differential_evolution(Three_hump_camel_fun, bounds, mutation=(0.1, 0.2), seed=42)
result4

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 813
     nit: 26
 success: True
       x: array([0., 0.])

Оптимум функции так же равен **0** <br>
Количество оценок целевых функций сократилось до **813**<br>
Количество выполненных итераций увеличелось до **26** <br>
Оптимум функции так же достигается в точке **[0, 0]**<br>

In [36]:
# Укажем, например, mutation=(1.9, 1.99)
result4 = differential_evolution(Three_hump_camel_fun, bounds, mutation=(1.9, 1.99), seed=42)
result4

     fun: 1.232595164407831e-30
 message: 'Optimization terminated successfully.'
    nfev: 13203
     nit: 439
 success: True
       x: array([0.00000000e+00, 1.11022302e-15])

Здесь уже точность немного пострадала: снова  в качестве решений получены значения близкие к нулю, но не равные нулю <br>
При этом <br>
Количество оценок целевых функций увеличелось до **13203**<br>
Количество выполненных итераций увеличелось до **439** <br>

In [40]:
# Теперь изменим параметр recombination - константа рекомбинации. По умолчанию recombination=0.7. 
# Укажем, например, recombination=0.99
result5 = differential_evolution(Three_hump_camel_fun, bounds, recombination=0.99, seed=42)
result5

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 2493
     nit: 82
 success: True
       x: array([0., 0.])

In [39]:
# Укажем, например, recombination=0.01
result5 = differential_evolution(Three_hump_camel_fun, bounds, recombination=0.01, seed=42)
result5

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 6303
     nit: 209
 success: True
       x: array([0., 0.])

В обоих случаях найдены верные решения. При увеличении значения константы рекомбинации количество целевых функций и количество выполненных итераций сокращается, а при уменьшении - увеличивается.

In [45]:
# Теперь изменим параметр updating. По умолчанию updating='immediate'. 
# Он может принимать второе значение updating='deferred'
result6 = differential_evolution(Three_hump_camel_fun, bounds, updating='deferred', seed=42)
result6

     fun: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 3783
     nit: 125
 success: True
       x: array([0., 0.])

Оптимум функции так же равен **0** <br>
Количество оценок целевых функций увеличелось до **3783**<br>
Количество выполненных итераций увеличелось до **125** <br>
Оптимум функции так же достигается в точке **[0, 0]**<br>

***

Постройте графики количества итераций оптимизации функции ackley от значения параметра.




In [ ]:
От значения какого параметра? Я могу любой параметр выбрать

В математической оптимизации функция  Ackley редставляет собой невыпуклую функцию, используемую в качестве проверки производительности алгоритмов оптимизации. Она была предложена Дэвидом Экли в его докторской диссертации 1987 года.

В двумерной области она определяется:

\begin{aligned}f(x,y)=-20&{}\exp \left[-0.2{\sqrt {0.5(x^{2}+y^{2})}}\,\right]\\&{}-\exp \left[0.5\left(\cos 2\pi x+\cos 2\pi y\right)\right]+e+20\end{aligned}
 
Ее глобальный оптимум:

 \begin{aligned}f(0,0)=0\end{aligned}